### PR TITLE
Add f3d/tex3d node, fix specular in SSS

### PR DESCRIPTION
**Pull request by luisbarrancos**
Since the optional token was incorrectly named in asSubsurface.
We also need a set of UDIM textures made on Mudbox, Zbrush, to test the atlas type in the texture() call.

### CODE REVIEW

File: src/appleseed.shaders/include/appleseed/texture/as_texture_helpers.h [hunk 1]
```diff
@@ -71,6 +71,26 @@ void set_wrap_mode(
     mode[1] = set_wrap_mode(wrap_tmode);
 }
 
+void set_wrap_mode(
+    int wrap_mode[3],
+    output string mode[3])
+{
+    mode[0] = set_wrap_mode(wrap_mode[0]);
```
Comment (dictoon) [inline src/appleseed.shaders/include/appleseed/texture/as_texture_helpers.h]: Something is wrong here. `set_wrap_mode()` doesn't return anything.

File: src/appleseed.shaders/src/appleseed/as_texture3d.osl [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        string as_maya_attribute_name = "filename",
+        string as_maya_attribute_short_name = "fil",
+        string label = "Filename",
+        string page = "3D Texture",
+        string widget = "filename",
+        int as_maya_attribute_connectable = 0,
+        int as_maya_attribute_keyable = 0,
+        int as_maya_attribute_hidden = 1,
+        int gafferNoduleLayoutVisible = 0,
+        string help = "The 3d texture file (field3d, f3d, etc)",
```
Comment (dictoon) [inline src/appleseed.shaders/src/appleseed/as_texture3d.osl]: *3D*

File: src/appleseed.shaders/src/appleseed/as_texture3d.osl [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        int gafferNoduleLayoutVisible = 0,
+        int divider = 1
+    ]],
+    color in_color = color(0)
+    [[
+        string as_maya_attribute_name = "color",
+        string as_maya_attribute_short_name = "c",
+        string label = "Default Color",
+        string page = "3D Texture",
+        string help = "Default Fill color to use if the texture lookup fails."
```
Comment (dictoon) [inline src/appleseed.shaders/src/appleseed/as_texture3d.osl]: *fill*

File: src/appleseed.shaders/src/appleseed/as_texture3d.osl [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+    ]],
+    float in_channel_fill = 0.0
+    [[
+        string as_maya_attribute_name = "channelFill",
+        string as_maya_attribute_short_name = "chf",
+        float softmin = 0.0,
+        float softmax = 1.0,
+        string label = "Default Channel Fill",
+        string page = "3D Texture",
+        string help = "Default Fill value for any channels requested but not present",
```
Comment (dictoon) [inline src/appleseed.shaders/src/appleseed/as_texture3d.osl]: *fill*

File: src/appleseed.shaders/include/appleseed/texture/as_texture_helpers.h (hunk 1 above)
Comment (luisbarrancos) [reply src/appleseed.shaders/include/appleseed/texture/as_texture_helpers.h]: There is a string return type set above. We could bypass this completely if we had the ability to associate a string value with another string (or any other value really) into the OSL token for the *mapper* options, but although there is apparently nothing forbidding it in the spec, it's not working in appleseed-maya so far.

Comment (dictoon) [reply src/appleseed.shaders/include/appleseed/texture/as_texture_helpers.h]: I see. I would suggest changing the name of the first `set_wrap_mode()` function (the one that returns a string): it doesn't set anything, instead it retrieves the string version of a wrap mode. Maybe `get_wrap_mode_name()`?

Comment (luisbarrancos) [reply src/appleseed.shaders/include/appleseed/texture/as_texture_helpers.h]: Good point, done.

**Pull request merged**